### DISCUSSION TITLE
Deleting cards in a page while switching between ui.tabs()

### DISCUSSION THREAD

**Discussion by n-srinidhi** [h2oai/wave#2028] [Q&A] [answered]
Hi! I'm trying to display different content separated into 2 tabs, and I alternate between them by deleting and adding respective cards on clicking/switching between different ui.tabs(). It worked for a few elements, However Now I am trying to add and delete a form_card(), and when I switch back and forth between the two tabs? The second time around I get the loading spinner page and nothing loads. Please refer to the code as given below:

This is how I'm initializing the page once while loading:

```py
q.page['most_graph']=ui.form_card(
            box=ui.box(zone='top_g',order=1,width='50%'),
            title='',
            items=[
                ui.inline(
                    items=[
                        ui.text_m(content='Most Popular Symbols'),
                        ui.dropdown(name='tog_most_pop',
                                choices=[ui.choice('tog_pop_o',label='Overall'),ui.choice('tog_pop_3',label='Last 3 Months'),ui.choice('tog_pop_today',label='Today')],
                                value='tog_pop_o',width='150px',trigger=True),
                    ],align='start',justify='between'
                ),
            ]
        )
```

Then in my serve function:

```py
elif q.args.header_tabs == 'menu_dash':
            q.page['header'].items[0].tabs.value='menu_dash'
            del q.page['similar_user']
            del q.page['bottom_section']
            q.page['most_graph']=ui.form_card(
                    box=ui.box(zone='top_g',order=1,width='50%'),
                    title='',
                    items=[
                        ui.inline(
                            items=[
                                ui.text_m(content='Most Popular Symbols'),
                                ui.dropdown(name='tog_most_pop',
                                        choices=[ui.choice('tog_pop_o',label='Overall'),ui.choice('tog_pop_3',label='Last 3 Months'),ui.choice('tog_pop_today',label='Today')],
                                        value='tog_pop_o',width='150px',trigger=True),
                            ],align='start',justify='between'
                        ),
                    ]
                )
elif q.args.header_tabs == 'menu_user':
            q.page['header'].items[0].tabs.value='menu_user'
            del q.page['most_graph']

meta=ui.meta_card(
    box='',
    title='Centra',
    layouts=[
            ui.layout(
            breakpoint='xl',
            zones=[
                ui.zone(name='header'),
                ui.zone(name='top'),
                ui.zone(name='top_g',direction=ui.ZoneDirection.ROW),
                ui.zone(name='middle'),
                ui.zone(name='bottom',direction=ui.ZoneDirection.ROW,zones=[
                    ui.zone('bot_left',size='50%'),
                    ui.zone('bot_right',size='50%')
                ]),
                ])
            ]
    ,
)
```

I understand the code might not be 100% reproducible, But I hope this is helpful in conveying the issue. What could be going wrong here?

**Comment by mturoci** [comment]
> and when I switch back and forth between the two tabs? The second time around I get the loading spinner page and nothing loads.

It's very hard to guess, but if you get the infinite loading spinner, it means you did not make any new UI changes before calling `q.page.save`. So if you delete the form card, you should replace it with something else. However, it's hard to tell without repro code.

**Reply by n-srinidhi** [reply]
I'm executing `del q.page['most_graph']` before calling `q.page.save()`

Basically, I need to display the `most_graph` form_card when I switch to the Dashboard Tab and need to delete it when I switch to User Tab. This is only happening when I have multiple cards in a single zone. I even tried splitting the zone like this:

```
ui.zone(name='top_g',direction=ui.ZoneDirection.ROW,zones=[
                    ui.zone('top_g_graph',size='50%'),
                    ui.zone('top_g_lst',size='50%')
                ]),
```
                
But got the same issue. This is happening only after the second or third to and fro switch between the `ui.tabs()`. Could it be some sort of an update issue?

**Reply by n-srinidhi** [reply, marked as answer]
@mturoci Hi! I fixed this. Found a bug in the code. Thanks!

**Reply by mturoci** [reply]
Great! I am glad you could figure it out :)
